proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
+3. Write out the parameters to the type class in order.
+4. The parameters must include a type defined in this file.
+5. Any generic type parameters must occur after the first instance of a local type or be covered by a local type.
+
+If a type class implementation meets these rules it is a valid type-side implementation.
+ 
+
+## Compile Resolution Rules
+
+When the compiler finds a call site invoking a function that has type class instances constraints declared with `with` as in the example below:
+
+Declaration: 
+```kotlin
+fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b)
+```
+Call site:
 ```kotlin
-extension object AddingInts {
-  fun addInts(): Int = add(1, 2)
+fun addInts(): Int = add(1, 2)
+```
+
+1. The compiler first looks at the file the interface is defined in. If it finds exactly one implementation it uses that instance.
+2. If it fails to find an implementation in the interface's file, it then looks at the files of the implemented types. For each type class parameter check the file it was defined in. If exactly one implementation is found use that instance.
+3. If no matching implementation is found in either of these places fail to compile.
+4. If more than one matching implementation is found, fail to compile and indicate that there or conflicting instances.
+
+Some of these examples where originally proposed by Roman Elizarov and the Kategory contributors where these features where originally discussed https://github.com/Kotlin/KEEP/pull/87
+
+## Appendix A: Orphan Implementations
+
+Orphan implementations are a subject of controversy. Combining two libraries, one defining a data type, the other defining an interface, is a feature that many programmers have longed for. However, implementing this feature in a way that doesn't break other features of interfaces is difficult and drastically complicates how the compiler works with those interfaces.
+
+Orphan implementations are the reason that type classes have often been described as "anti-modular" as the most common way of dealing with them is through global coherency checks. This is necessary to ensure that two libraries have not defined incompatible implementations of a type class interface.
+
+Relaxing the orphan rules is a backwards-compatible change. If this proposal is accepted without permitting orphans it is useful to consider how they could be added in the future.
+
+Ideally we want to ban orphan implementations in libraries but not in executables; this allows a programmer to manually deal with coherence in their own code but prevents situations where adding a new library breaks code.
+
+### Package-based Approach to Orphans
+
+A simple way to allow orphan implementations is to replace the file-based restrictions with package-based restrictions. Because there are no restrictions on packages it is posible to do the following.
```

Comment:
I recall someone telling me that redefining packages had restriction or issues after JDK 9, not sure though.

Reply:
This is one of the reasons that I think it best to wait on the orphan question. Java 9 modules will really change how libraries work on the JVM and that would really impact something like type classes.